pkg/utils/meta_loader.go

```diff
@@ -0,0 +1,134 @@
+		Env:       env,
+		Version:   version,
+	}
+}
+
+// Load - loads the addon metadata and imageSet
+func (l defaultMetaLoader) Load() (*addonsv1alpha1.AddonMetadataSpec, error) {
+	meta, err := l.readMeta()
+	if err != nil {
+		return nil, err
+	}
+	// invalid - legacy addon
+	if meta.IndexImage == nil && meta.ImageSetVersion == nil {
+		return nil, errors.New("No validation support for legacy addon. Please use the imageSet feature.")
+	}
+	// invalid - misconfiguration
+	if meta.IndexImage != nil && meta.ImageSetVersion != nil {
+		return nil, errors.New("Can't set both the 'indexImage' and the 'imageSetVersion' field.")
+	}
+	// imageSet
+	if meta.ImageSetVersion != nil {
+		imageSet, err := l.readImageSet(*meta.ImageSetVersion)
+		if err != nil {
+			return nil, fmt.Errorf("Could not read imageSet, got %v.\n", err)
+		}
+		// deepcopy to be safe
+		meta.IndexImage = &imageSet.IndexImage
+		meta.AddOnParameters = imageSet.AddOnParameters.DeepCopy()
+		meta.AddOnRequirements = imageSet.AddOnRequirements.DeepCopy()
+		meta.SubOperators = imageSet.SubOperators.DeepCopy()
+	}
+	return meta, nil
+}
+
+func (l defaultMetaLoader) readMeta() (*addonsv1alpha1.AddonMetadataSpec, error) {
+	data, err := ioutil.ReadFile(l.getMetadataPath())
+	if err != nil {
+		return nil, err
+	}
+	log.Debugf("Raw metadata read from addon: %v. \n%v\n", l.AddonName, string(data))
+	meta := &addonsv1alpha1.AddonMetadataSpec{}
+	err = meta.FromYAML(data)
+	return meta, err
+}
+
+func (l defaultMetaLoader) getMetadataPath() string {
+	return path.Join(l.AddonDir, "metadata", l.Env, "addon.yaml")
+}
+
+func (l defaultMetaLoader) readImageSet(defaultVersion string) (*addonsv1alpha1.AddonImageSetSpec, error) {
+	version := l.getImageSetVersion(defaultVersion)
+	imageSetPath, err := l.getImagesetPath(version)
+	if err != nil {
+		return nil, err
+	}
+	data, err := ioutil.ReadFile(imageSetPath)
+	if err != nil {
+		return nil, err
+	}
+	log.Debugf("Raw imageSet read from addon: %v. \n%v\n", l.AddonName, string(data))
+	imageSet := &addonsv1alpha1.AddonImageSetSpec{}
+	err = imageSet.FromYAML(data)
+	return imageSet, err
+}
+
+// defaultVersion == meta.ImageSetVersion
+// Can be overriden by providing the --version CLI flag
+func (l defaultMetaLoader) getImageSetVersion(defaultVersion string) string {
+	if l.Version != "" {
+		return l.Version
+	}
+	return defaultVersion
+}
+
+func (l defaultMetaLoader) getImagesetPath(version string) (string, error) {
+	baseDir := path.Join(l.AddonDir, "addonimagesets", l.Env)
+	target := fmt.Sprintf("%s.%s.yaml", l.AddonName, version)
+	if version == "latest" {
+		latest, err := getLatestImageSetVersion(baseDir)
+		if err != nil {
+			return "", err
+		}
+		target = latest
+	}
+	return path.Join(baseDir, target), nil
+}
+
+func getLatestImageSetVersion(dir string) (string, error) {
+	sortDescending := func(files []os.FileInfo) {
+		sort.Slice(files, func(i, j int) bool {
+			return files[i].Name() > files[j].Name()
```

Comment:
To have parity with what we have in the python package, can we read the contents `Name` from the file and then sort accordingly?

Reply:
Is this necessary? If the filename format is enforced as `<name>.v<version>.yaml` don't we get the same information?

Reply:
Yep, we'll add a validation to enforce the filename.